tests/unit/APNSClientTest.php

```diff
@@ -0,0 +1,30 @@
+<?php
+declare( strict_types = 1 );
+
+class APNSClientTest extends APNSTest {
+
+	public function testThatRequestsFromNetworkAreProcessed() {
+		$network_service_mock = Mockery::mock( 'MultiplexedNetworkService' );
+		$network_service_mock
+			->shouldReceive( 'enqueueRequest' )
+			->once();
+		$fake_responses = [
+			new Response( 200, $this->new_apns_http_failure_response( 200 ), 1, 1 ),
+			new Response( 400, $this->new_apns_http_failure_response( 400 ), 1, 1 ),
+		];
+		$network_service_mock
+			->shouldReceive( 'sendQueuedRequests' )
+			->andReturn( $fake_responses );
+		$client = new APNSClient( $this->new_configuration(), $network_service_mock );
+
+		$responses = $client->sendRequests( [ $this->new_request() ] );
+
+		$this->assertEquals( 2, count( $responses ) );
+		// For the purpose of this test it's enough to verify that the received
+		// requests are mapped to the expected type by means of calling a
+		// method on them. It's then up to the test for the type initialization
+		// to verify the mapping is done correctly.
+		$this->assertEquals( 200, $responses[0]->getStatusCode() );
+		$this->assertEquals( 400, $responses[1]->getStatusCode() );
+	}
```

Comment:
This test is what all that work lead up to. It's more at the integration side of things, and because of how the mocking framework works (or likely my limited familiarity with it) it feels like there's a bit too much implementation detail leaking into it.

@jkmassel what do you think of this? Does it seem useful or, at least, promising?

Reply:
I think this test is great – it allows us to test the `APNSClient` class with relatively minimal boilerplate.

What abstraction leaks do you see in the test? Because it seems fairly clean to me! 😃

Reply:
Thanks.

Oh, nothing much really. I just don't like all the calls with strings for the method names. They reveal too much of what's going on under the hood in the networking layer and are also brittle, as soon as a method name changes they'll need updating.

But, it is what it is I guess. That's what using test doubles involves.